flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -773,6 +773,13 @@ static boolean shouldSyncAfterClosingLogicalFile(FileSystem fileSystem) {
         return true;
     }
 
+    private static String uriEscape(String input) {
+        // All reserved characters (RFC 2396) will be removed. This is enough for flink's resource
+        // id, job id and operator id.
+        // Ref: https://docs.oracle.com/javase/8/docs/api/index.html?java/net/URI.html
+        return input.replaceAll("[;/?:@&=+$,\\[\\]]", "-");
+    }
```

Comment:
It seems work after this PR. I still have 2 questions:

1. Do flink or java have the utility class to remove reserved characters before? 
    - I guess not only `FileMerging` feature needs to consider it. So I'm not sure should the `uriEscape` be put here.
2. For test, did you try run a real job locally with this PR? I'm not sure is there other issues if we run `FileMerging` feature on a standalone cluster.

Reply:
1. I did some investigation. There are some utilities for URI from third-party dependencies, but the behavior is not same as what we want. Instead of removing or replacing the reserved characters, they may treat the reserved characters as valid ones. The `java.net.URLEncoding` seems feasible, however, URL specification is a little bit different from the URI. So I'd prefer doing a simple replace, since the ids from flink is neat.
2. Well no, we never did that. We tried Per-Job/Session on yarn and local mini-cluster (E2E). I'm trying the local standalone cluster.